tools/spectral/ipa/rulesets/functions/IPA125OneOfMustHaveDiscriminator.js

```diff
@@ -0,0 +1,50 @@
+import { collectAdoption, collectAndReturnViolation } from './utils/collectionUtils.js';
+import { resolveObject } from './utils/componentUtils.js';
+import { hasException } from './utils/exceptions.js';
+
+const RULE_NAME = 'xgen-IPA-125-oneOf-must-have-discriminator';
+const ERROR_MESSAGE = 'Each oneOf property must include a discriminator property to define the exact type.';
+const MAPPING_ERROR_MESSAGE =
+  'The discriminator mapping must match the oneOf references. Unmatched Discriminator mappings with oneOf references:';
+
+export default (input, _, { path, documentInventory }) => {
+  const oas = documentInventory.unresolved; // Use unresolved document to access raw $ref
+  const schema = resolveObject(oas, path);
+
+  if (!schema || !schema.oneOf || !Array.isArray(schema.oneOf)) {
+    return;
+  }
+
+  // Check for exception first
+  if (hasException(schema, RULE_NAME)) {
+    return;
+  }
+
+  // Check if all oneOf items are objects with a $ref property
+  const allReferences = schema.oneOf.every((item) => typeof item === 'object' && item.$ref);
+  if (!allReferences) {
+    return;
+  }
+
+  // Validate the presence of a discriminator
+  if (!schema.discriminator || typeof schema.discriminator !== 'object' || !schema.discriminator.propertyName) {
+    return collectAndReturnViolation(path, RULE_NAME, [{ message: ERROR_MESSAGE }]);
```

Comment:
There might be a different error message for different cases such as

```suggestion
if(!schema.discriminator)
  message = "The schema has oneOf but not discriminator property"
if(typeof schema.discriminator !== 'object')
  message = "Discriminator property is not an object"
if (!schema.discriminator.propertyName)
   message = "Discriminator has no propertyName defined"  
```

Messages are not their best, but tried to show my point

Reply:
🙇